File: ruby/test/05_fake_test.rb
# 👩‍🏫 05_fake_test missions:
#    - Implement the FakeRepository in order to make the tests pass. Take a look at CrudApp in lib/05_fake for API details.
#    - If you're struggling, an in-memory data store using a HashMap/object is probably the most straightforward route.

require "minitest/autorun"
require "05_fake"

# ✍️ WRITE YOUR TEST DOUBLE HERE

class FakeRepository
end

# ✍️ END

# 🙅‍♀️ DO NOT CHANGE THIS CODE - START

class TestCrudApp < Minitest::Test
  def test_can_create_and_read
    # Arrange
    repository = FakeRepository.new
    subject = CrudApp.new(repository)
    username = "Ash"
    favourite_pokemon = "Pikachu"

    # Act
    id = subject.create(username, favourite_pokemon)
    result = subject.read(id)

    # Assert
    assert_equal username, result[:username]
    assert_equal favourite_pokemon, result[:favourite_pokemon]
  end

  def test_can_update
    # Arrange
    repository = FakeRepository.new
    subject = CrudApp.new(repository)
    username = "Ash"
    old_favourite = "Pikachu"
    new_favourite = "Charizard"

    # Act
    id = subject.create(username, old_favourite)
    updated = subject.update(id, username, new_favourite)
    result = subject.read(id)

    # Assert
    assert updated
    assert_equal username, result[:username]
    assert_equal new_favourite, result[:new_favourite]
  end

  def test_will_raise_if_id_is_deleted
    # Arrange
    repository = FakeRepository.new
    subject = CrudApp.new(repository)
    username = "Ash"
    favourite_pokemon = "Pikachu"

    # Act
    id = subject.create(username, favourite_pokemon)
    deleted = subject.delete(id)

    # Assert
    assert deleted
    assert_raises { subject.read(id) }
  end
end

# 🙅‍♀️ DO NOT CHANGE THIS CODE - END

File: javascript/src/05_fake.test.js
const CrudApp = require("./05_fake");

class FakeRepository {}

describe("CrudApp, Fake Test Double", () => {
  test("it can create and read", () => {
    // Arrange
    const repository = new FakeRepository();
    const subject = new CrudApp();
    const username = "Ash";
    const favouritePokemon = "Pikachu";

    // Act
    const id = subject.create(username, favouritePokemon);
    const result = subject.read(id);

    // Assert
    expect(result.username).toEqual(username);
    expect(result.favoritePokemon).toEqual(favouritePokemon);
  });

  test("it can update", () => {
    // Arrange
    const repository = new FakeRepository();
    const subject = new CrudApp();
    const username = "Ash";
    const oldFavourite = "Pikachu";
    const newFavourite = "Charizard";

    // Act
    const id = subject.create(username, oldFavourite);
    const updated = subject.update(id, username, newFavourite);
    const result = subject.read(id);

    // Assert
    expect(updated).not.toBeNull();
    expect(result.username).toEqual(username);
    expect(result.favoritePokemon).toEqual(newFavourite);
  });

  test("will raise if id is deleted", () => {
    // Arrange
    const repository = new FakeRepository();
    const subject = new CrudApp();
    const username = "Ash";
    const favouritePokemon = "Pikachu";

    // Act
    const id = subject.create(username, favouritePokemon);
    const deleted = subject.delete(id);

    // Assert
    expect(deleted).not.toBeNull();
    expect(() => {
      subject.read(id);
    }).toThrow();
  });
});
